File: im-ai-office/imai/services/tasks.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""任务状态流转服务（自 core.py:443-503 1:1 迁移）：confirm/reject/list/update"""
import json
from datetime import datetime

from imai.config import EVENTS
from imai.repos import audit_log

CANCELLED = "cancelled"


def list_tasks(con, status=None):
    from imai.repos import list_task_dicts
    return list_task_dicts(con, status)


def confirm_task(con, task_id, assignee=None, deadline=None):
    c = con.cursor()
    if assignee:
        c.execute("UPDATE task SET status='confirmed', assignee=?, updated_at=datetime('now') WHERE id=?",
                  (assignee, task_id))
    else:
        c.execute("UPDATE task SET status='confirmed', updated_at=datetime('now') WHERE id=?", (task_id,))
    con.commit()
    c.execute("SELECT * FROM task WHERE id=?", (task_id,))
    row = c.fetchone()
    if row:
        # 到期提醒（简化：临近截止提醒）
        if row["deadline"] and ("周五" in row["deadline"] or "明天" in row["deadline"] or "天内" in row["deadline"]):
            EVENTS.append({"event": "reminder.due", "taskId": task_id, "assignee": row["assignee"], "tier": "due"})
    c.execute("INSERT INTO audit(actor,action,detail) VALUES('user','confirm',?)",
              (json.dumps({"taskId": task_id}, ensure_ascii=False),))
    con.commit()
    return True


def complete_task(con, task_id, actor="user"):
    """G1 完成回流（工作流缺口登记与完成回流Spec §1）：confirmed/pending → done。

    done 是新增终态：提醒扫描白名单不含 done → 逾期提醒自然终止。"""
    c = con.cursor()
    c.execute("SELECT status FROM task WHERE id=?", (task_id,))
    row = c.fetchone()
    if not row:
        return False
    if (row["status"] or "") not in ("confirmed", "pending_confirmation", "pending_assignee"):
        return False
    c.execute("UPDATE task SET status='done', updated_at=datetime('now') WHERE id=?", (task_id,))
    con.commit()
    audit_log(con, actor, "task_completed", {"taskId": task_id})
    EVENTS.append({"event": "task.completed", "taskId": task_id})
    return True


def update_task(con, task_id, assignee=None, deadline=None, cancel=False):
    """迭代2 B1：已确认任务修改（改负责人/改期/取消）。

    返回 (row, err)；err ∈ task_not_found / bad_deadline / no_changes / None。
    deadline 变更时清空该任务 reminder_sent（三档提醒按新时间重新起算，Spec §1.2）。
    每处变更写 audit(action='task_update', detail={field, old, new})。
    """
    c = con.cursor()
    c.execute("SELECT * FROM task WHERE id=?", (task_id,))
    row = c.fetchone()
    if not row:
        return None, "task_not_found"
    changes = {}
    if assignee is not None and assignee != row["assignee"]:
        c.execute("UPDATE task SET assignee=?, updated_at=datetime('now') WHERE id=?",
                  (assignee, task_id))
        changes["assignee"] = (row["assignee"], assignee)
    if deadline is not None:
        try:
            dt = datetime.strptime(deadline, "%Y-%m-%d %H:%M")
        except ValueError:
            return None, "bad_deadline"
        c.execute("UPDATE task SET deadline=?, deadline_at=?, updated_at=datetime('now') WHERE id=?",
                  (deadline, dt.strftime("%Y-%m-%d %H:%M"), task_id))
        changes["deadline"] = (row["deadline"], deadline)
        c.execute("DELETE FROM reminder_sent WHERE task_id=?", (task_id,))
    if cancel:
        c.execute("UPDATE task SET status=?, updated_at=datetime('now') WHERE id=?",
                  (CANCELLED, task_id))
        changes["status"] = (row["status"], CANCELLED)
    con.commit()
    if not changes:
        return None, "no_changes"
    for field, (old, new) in changes.items():
        audit_log(con, "user", "task_update",
                  {"taskId": task_id, "field": field, "old": old, "new": new})
        if field == "deadline":
            audit_log(con, "user", "reminder_reset", {"taskId": task_id})
    c.execute("SELECT * FROM task WHERE id=?", (task_id,))
    return c.fetchone(), None


def reject_task(con, task_id, reason="", assignee=None):
    c = con.cursor()
    c.execute("UPDATE task SET status='rejected', updated_at=datetime('now') WHERE id=?", (task_id,))
    con.commit()
    audit_log(con, "user", "reject", {"taskId": task_id, "reason": reason})
    # S4/M4: 修正信号沉淀 —— 若驳回理由指明正确负责人，更新人称别名
    from imai.services.memory import _memorize_reject_signal
    _memorize_reject_signal(con, reason, task_id)
    return True

File: im-ai-office/backend-ts/src/tasks.ts
import { and, eq, inArray, sql } from "drizzle-orm";
import { db } from "./db/drizzle.js";
import { reminderSent, task } from "./db/schema.js";
import { auditLog, getTaskDict, type TaskRow } from "./repos.js";
import { fanout } from "./sse.js";
import { UNRESOLVED_STATUS } from "./config.js";

// ============ 任务状态流转（tasks.py 的 TS 版）：confirm/reject/complete/update ============
// P1 加固：状态流转一律条件更新（WHERE status IN ...）+ 事务，杜绝
//   - 并发双 confirm/双 complete 同时成功（check-then-act 竞态）
//   - updateTask 半改状态（原：先改 assignee 再校验 deadline，失败即留半截）
// 审计写在事务外（best-effort 留痕）：状态流转的原子性是底线，审计失败不回滚状态。

export const CANCELLED = "cancelled";

const touch = { updatedAt: sql`NOW()` };

export async function confirmTask(taskId: number, assignee?: string | null, _deadline?: string | null, actor = "user"): Promise<boolean> {
  const ok = await db.transaction(async (tx) => {
    const vals: Record<string, unknown> = { status: "confirmed", ...touch };
    if (assignee !== undefined && assignee !== null) vals.assignee = assignee;
    const rows = await tx.update(task).set(vals)
      .where(and(eq(task.id, taskId), inArray(task.status, [...UNRESOLVED_STATUS])))
      .returning({ id: task.id });
    return rows.length > 0;
  });
  if (!ok) return false;
  await auditLog(actor, "confirm", { taskId });
  fanout("task_status", { taskId, status: "confirmed" });
  return true;
}

export async function rejectTask(taskId: number, reason?: string | null, actor = "user"): Promise<boolean> {
  const ok = await db.transaction(async (tx) => {
    const rows = await tx.update(task).set({ status: "rejected", ...touch })
      .where(and(eq(task.id, taskId), inArray(task.status, [...UNRESOLVED_STATUS, "confirmed"])))
      .returning({ id: task.id });
    return rows.length > 0;
  });
  if (!ok) return false;
  await auditLog(actor, "reject", { taskId, reason: reason ?? "" });
  // S4/M4：修正信号沉淀——驳回理由指明正确负责人时，更新人称别名
  if (reason) {
    const { memorizeRejectSignal } = await import("./memory.js");
    await memorizeRejectSignal(reason, taskId);
  }
  fanout("task_status", { taskId, status: "rejected" });
  return true;
}

/** G1 完成回流：confirmed/pending → done（提醒扫描白名单不含 done，逾期提醒自然终止）。 */
export async function completeTask(taskId: number, actor = "user"): Promise<boolean> {
  const rows = await db.update(task).set({ status: "done", ...touch })
    .where(and(eq(task.id, taskId), inArray(task.status, ["confirmed", ...UNRESOLVED_STATUS])))
    .returning({ id: task.id });
  if (!rows.length) return false;
  await auditLog(actor, "task_completed", { taskId });
  return true;
}

/** 迭代2 B1：已确认任务修改（改负责人/改期/取消）。返回 {task?, err?}，语义 1:1 对齐 Python。 */
export async function updateTask(
  taskId: number, assignee?: string | null, deadline?: string | null, cancel = false, actor = "user"
): Promise<{ task?: TaskRow; err?: string }> {
  const row = await getTaskDict(taskId);
  if (!row) return { err: "task_not_found" };
  // 校验先行：坏 deadline 在任何写之前返回，不再留下"assignee 已改、deadline 报错"的半改状态
  if (deadline !== undefined && deadline !== null && !/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}$/.test(deadline)) {
    return { err: "bad_deadline" };
  }
  const changes: Record<string, unknown> = {};
  await db.transaction(async (tx) => {
    if (assignee !== undefined && assignee !== null && assignee !== row.assignee) {
      await tx.update(task).set({ assignee, ...touch }).where(eq(task.id, taskId));
      changes["assignee"] = [row.assignee, assignee];
    }
    if (deadline !== undefined && deadline !== null) {
      await tx.update(task).set({ deadline, deadlineAt: deadline, ...touch }).where(eq(task.id, taskId));
      changes["deadline"] = [row.deadline, deadline];
      await tx.delete(reminderSent).where(eq(reminderSent.taskId, taskId));
    }
    if (cancel) {
      await tx.update(task).set({ status: CANCELLED, ...touch }).where(eq(task.id, taskId));
      changes["status"] = [row.status, CANCELLED];
    }
  });
  await auditLog(actor, "task_update", { taskId, changes });
  const updated = await getTaskDict(taskId);
  fanout("task_status", { taskId, status: updated!.status ?? "" });
  return { task: updated! };
}
